src/struct_metadata.rs

```diff
@@ -0,0 +1,166 @@
+// FIXME: this is allowed only temporarily,
+//  before the insert RBWNAT implementation is ready,
+//  cause otherwise the caches are never used.
+#![allow(dead_code)]
+#![allow(unreachable_pub)]
+
+use crate::row::RowType;
+use crate::sql::Identifier;
+use crate::Result;
+use crate::Row;
+use clickhouse_types::{parse_rbwnat_columns_header, Column};
+use std::collections::HashMap;
+use std::fmt::Display;
+use std::sync::Arc;
+use tokio::sync::{OnceCell, RwLock};
+
+/// Cache for [`StructMetadata`] to avoid allocating it for the same struct more than once
+/// during the application lifecycle. Key: fully qualified table name (e.g. `database.table`).
+type LockedStructMetadataCache = RwLock<HashMap<String, Arc<StructMetadata>>>;
+static STRUCT_METADATA_CACHE: OnceCell<LockedStructMetadataCache> = OnceCell::const_new();
+
+#[derive(Debug, PartialEq)]
+enum AccessType {
+    WithSeqAccess,
+    WithMapAccess(Vec<usize>),
+}
+
+/// [`StructMetadata`] should be owned outside the (de)serializer,
+/// as it is calculated only once per struct. It does not have lifetimes,
+/// so it does not introduce a breaking change to [`crate::cursors::RowCursor`].
+pub struct StructMetadata {
+    /// See [`Row::NAME`]
+    pub(crate) struct_name: &'static str,
+    /// See [`Row::COLUMN_NAMES`] (currently unused)
+    pub(crate) struct_fields: &'static [&'static str],
+    /// See [`Row::TYPE`]
+    pub(crate) row_type: RowType,
+    /// Database schema, or columns, are parsed before the first call to (de)serializer.
+    pub(crate) columns: Vec<Column>,
+    /// This determines whether we can just use [`crate::rowbinary::de::RowBinarySeqAccess`]
+    /// or a more sophisticated approach with [`crate::rowbinary::de::RowBinaryStructAsMapAccess`]
+    /// to support structs defined with different fields order than in the schema.
+    /// (De)serializing a struct as a map will be approximately 40% slower than as a sequence.
+    access_type: AccessType,
+}
+
+impl StructMetadata {
+    // FIXME: perhaps it should not be public? But it is required for mocks/provide.
```

Comment:
It shouldn't; I would like to disable checks for mocks because they are meaningless there.

To do it, we should either check some special header if the `test-utils` feature is enabled or change the api from
```
    let client = Client::default().with_url(mock.url());
```
to something like
```
    let client = Client::default().with_mock(&mock);
```
and remove `Mock::url()` totally

Reply:
It can be used as:

```rs
let client = Client::default()
    .with_url(mock.url())
    .with_disabled_validation();
```

Since after https://github.com/ClickHouse/clickhouse-rs/pull/221/commits/ccfac33235d978716902884effe3ffd8c0261b3d we will not allow validating only first N rows, it is now either enabled or disabled.

We can add a different API for that, but `with_disabled_validation` is just faster to implement.